shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -117,15 +191,56 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
   result(handled ? nil : FlutterMethodNotImplemented);
 }
 
+/**
+ * Updates the text input model with state received from the framework via the
+ * TextInput.setEditingState message.
+ */
+- (void)setEditingState:(NSDictionary*)state {
+  NSString* selectionAffinity = state[kSelectionAffinityKey];
+  if (selectionAffinity != nil) {
+    _textAffinity = [selectionAffinity isEqualToString:kTextAffinityUpstream]
+                        ? FlutterTextAffinityUpstream
+                        : FlutterTextAffinityDownstream;
+  }
+
+  NSString* text = state[kTextKey];
+  if (text != nil) {
+    _active_model->SetText([text cStringUsingEncoding:NSUTF8StringEncoding]);
+  }
+
+  flutter::TextRange selected_range = UpdateRangeFromBaseExtent(
+      state[kSelectionBaseKey], state[kSelectionExtentKey], _active_model->selection());
+  _active_model->SetSelection(selected_range);
+
+  flutter::TextRange composing_range = UpdateRangeFromBaseExtent(
+      state[kComposingBaseKey], state[kComposingExtentKey], _active_model->composing_range());
+  size_t cursor_offset = selected_range.base() - composing_range.start();
+  _active_model->SetComposingRange(composing_range, cursor_offset);
+}
+
 /**
  * Informs the Flutter framework of changes to the text input model's state.
  */
 - (void)updateEditState {
-  if (self.activeModel == nil) {
+  if (_active_model == nullptr) {
     return;
   }
-  [_channel invokeMethod:kUpdateEditStateResponseMethod
-               arguments:@[ self.activeModel.clientID, self.activeModel.state ]];
+
+  NSString* const textAffinity = (self.textAffinity == FlutterTextAffinityUpstream)
+                                     ? kTextAffinityUpstream
+                                     : kTextAffinityDownstream;
+  NSDictionary* state = @{
+    kSelectionBaseKey : @(_active_model->selection().base()),
+    kSelectionExtentKey : @(_active_model->selection().extent()),
+    kSelectionAffinityKey : textAffinity,
+    kSelectionIsDirectionalKey : @NO,
+    kComposingBaseKey : @(_active_model->composing_range().base()),
+    kComposingExtentKey : @(_active_model->composing_range().extent()),
+    kTextKey : [NSString stringWithCString:_active_model->GetText().c_str()
+                                  encoding:NSUTF8StringEncoding]
```

Comment:
Similarly this can be spelled `stringWithUTF8String:`

Reply:
And done.